### ISSUE TITLE
A Kit with toolchainFile fails to determine CMAKE_(C|CXX)_COMPILER without the user adding both variables to the cache file

### ISSUE THREAD

**Issue by edgomez** [microsoft/vscode-cmake-tools#1188]
### Brief Issue Summary

A local user CMake kit using a toolchainFile will make cmake-tools fail at determining the C/C++ compiler path.

Then IntelliSense setup is done wrong:
1. Architecture/Pltform/CompilerFlavor dependent codepaths highlighting is wrong (eg:  `__amd64__` or `__arm__`, `_WIN32` or `__linux__`, `__MINGW32__` or `__MINGW64__`)
2. Code completion is made difficult or incorrect (because of 1., symbols may be missing etc...)

### Expected:

CMake [toolchain file documentation](https://cmake.org/cmake/help/latest/manual/cmake-toolchains.7.html) doesn't mandate that `CMAKE_C_COMPILER` and `CMAKE_CXX_COMPILER`  variables be placed in the cache. 

`CMAKE_C_COMPILER` and `CMAKE_CXX_COMPILER` should be figured out automatically by the extension w/o adding a cache entry for them both.

Note that the cache workaround is mentioned in:
- #602 where writing the CXX/C compiler settings forces them on the command line and thus in the cache
- #500 where a specific pair of `set(... CACHE ...)` calls are inserted in the toolchain file

### Reproduction steps:

1. Setup a local CMake build kit using a toolchain file documented as per CMake guidelines. (eg: a buildroot generated toolchain file will do just fine)
2. Import a C++ project for which code blocks vary depending on the architecture (eg: `__amd64__` or `__arm__`)  or on the platform (eg: `_WIN32` or `__linux__`) or the compiler flavor (eg: `__MINGW32__` or `__MINGW64__`)
3. `Ctrl+Shift+P` `Cmake: Configure`
4. Observe the codepaths highlighting mismatches and the code completion problems

### Apparent Behavior:

CMake Tools expects `CMAKE_C_COMPILER` and `CMAKE_CXX_COMPILER` in the cache to setup the CPP project correctly.

### CMake Tools Log

No relevant logs to copy/paste.

### Developer Tools Log

No relevant logs to copy/paste.

### Platform and Versions

- **Operating System**: Debian GNU/Linux SID
- **CMake Version**: 3.16.3
- **VSCode Version**: Version: 1.44.2 Commit: ff915844119ce9485abfe8aa9076ec76b5300ddd
- **CMake Tools Extension Version**: 1.3.1
- **Compiler/Toolchain**:  any cross toolchain would exhibit the problem as long as it's setup only through a toolchain file. (tested with a GCC 7.3 generated by a ARM based buildroot build, or Debian's (g++|gcc)-arm-linux-gnueabi packages, or Debian's gcc-mingw-w64-(x86-64|i686)) 

### Other Notes/Information

Here is a failing toolchain file example. It targets mingw32 builds using mingw32 debian packages.

```cmake
set(CMAKE_SYSTEM_NAME Windows)
set(CMAKE_SYSTEM_PROCESSOR x86)

set(toolchain_dir "/usr/bin")
set(toolchain_triplet "i686-w64-mingw32")
set(toolchain_prefix "${toolchain_triplet}-")
set(toolchain_suffix "-posix")

set(CMAKE_SYSROOT "/usr/${toolchain_triplet}")

set(CMAKE_C_COMPILER   "${toolchain_dir}/${toolchain_prefix}gcc${toolchain_suffix}")
set(CMAKE_CXX_COMPILER "${toolchain_dir}/${toolchain_prefix}g++${toolchain_suffix}")
set(CMAKE_RC_COMPILER "${toolchain_dir}/${toolchain_prefix}windres" CACHE FILEPATH "" FORCE)

set(CMAKE_FIND_ROOT_PATH_MODE_PROGRAM NEVER)
set(CMAKE_FIND_ROOT_PATH_MODE_LIBRARY ONLY)
set(CMAKE_FIND_ROOT_PATH_MODE_INCLUDE ONLY)
set(CMAKE_FIND_ROOT_PATH_MODE_PACKAGE BOTH)
```

Appending these two lines are the current workaround for hand made toolchain files
```cmake
set(CMAKE_C_COMPILER "${CMAKE_C_COMPILER}" CACHE FILEPATH "C Compiler path")
set(CMAKE_CXX_COMPILER "${CMAKE_CXX_COMPILER}" CACHE FILEPATH "C++ Compiler path")
```

The problem can't be workaround easily when using generated toolchain files (eg: buildroot ./host/usr/share/buildroot/toolchainfile.cmake) or external toolchain files provided by third parties.

**Comment by andreeis** [comment]
@edgomez, how about @bobbrow's comment at the end of issue https://github.com/microsoft/vscode-cmake-tools/issues/602, that didn't work for you?

Copy pasting the suggestion:
"You should be able to set the C and CXX compilers without editing the Cache by adding a "Kit" to your user-local or project kits file: https://vector-of-bool.github.io/docs/vscode-cmake-tools/kits.html#kits.
Then select that Kit and reconfigure the project."

**Comment by edgomez** [comment]
@andreeis, indeed this actually works.

But, IMHO this still qualifies as a workaround because vscode-cmake-tools need those variables cached to configure correctly the CPP backend.

Here is my reasoning:
1. According to cmake-toolchains(7) documentation, having the C/CXX compiler defined as simple variables in the toolchain file should suffice. And all toolchain files I came across and those I write myself do use regular set(CMAKE_C|CXX_COMPILER value), no CACHE involved.
2.  cmake-toolchains(7) command line examples tend to suggest that one can complement some toolchain files w/ additional variables directly defined on the command line, this is for example mentioned for Android or iOS support. But no mention of the necessity to define CMAKE_C|CXX_compiler using -D options on top of the toolchain file itself.
3. vscode-cmake-tools documentation tells configuring a local kit is as simple as filling in the C/CXX properties. Then the only example found for the toolchainFile property, totally omits to fill the C/CXX properties. This strongly suggests you either fill in the C/CXX properties, or you fill in the toolchainFIle property. No need for both.
4. Additionally, vscode-cmake-tools seems to use the `compile_commands.json` file to grab lots of information for the CPP backend (command line defines etc...). It could find the compilers used in each "command" entry's first token.

So to me, the fact we need to fill in the C/CXX properties in a CMakeKit defeats the whole purpose to fill all the required information in a toolchain file to support a different target system:
* If this is not considered a bug/vscode-cmake-tools shortcoming, then the documentation could be at least rephrased to make clearer that if one wants perfect CPP backend configuration by the extension, the C/CXX must be filled in even if the `toolchainFile` property is already used and the file defines those two.
* If current behavior is considered a shortcoming of the current vscode-cmake-tools implementation, then C/CXX could be deduced differently. From the `compile_commands.json` file then provided to the CPP backend the same way command line defines are found. Or by asking nicely cmake to output the C/CXX vars itself and fill automatically those when needed in the CMakeKit. This would avoid users setting up the properties themselves and potentially create configuration mistakes (what happens if the toolchainFile and C/CXX properties don't agree on which compiler is "used", the CPP backend will still be mis-configured). 

whats is your opinion ?

**Comment by edgomez** [comment]
To illustrate the proposal in my comment.

When configuring a cmake kit, vscode-cmake-tools could configure a temporary project whose sole CMakeList.txt would provide it with the C/CXX compiler information it needs. This can be done as follow:

```cmake
cmake_minimum_required(VERSION 2.8)

project(PrintCLanguageCompilersForVSCode)

enable_language(C CXX)

message(STATUS "vscode-cmake-tools kit configuration: CMAKE_C_COMPILER=${CMAKE_C_COMPILER}")
message(STATUS "vscode-cmake-tools kit configuration: CMAKE_CXX_COMPILER=${CMAKE_CXX_COMPILER}")
```
The result could be grep'ed from cmake output, or maybe communicated over whatever the extension cmake driver supports.

What do you think ?

**Comment by andreeis** [comment]
@edgomez, makes sense. We need to digest this proposal more. Also, upvoting the first comment in this issue will help us to prioritize it appropriately.

**Comment by edgomez** [comment]
Ok up-voted my first comment. Hope this is picked for a future release soon :-)
If you need any further detail, do not hesitate to ask.

**Comment by aelray** [comment]
I think I have a similar issue. We're using the proprietary ARM compiler (ARMClang), and here even the workaround putting the compiler in the cache or setting with the cmake-tools-kits.json does not help to make the cpp code highlighting work correctly.

`/*
 * ARM Compiler 6 (armclang)
 */
#elif defined (__ARMCC_VERSION) && (__ARMCC_VERSION >= 6010050)
  #include "cmsis_armclang.h"


/*
 * GNU Compiler
 */
#elif defined ( __GNUC__ )
  #include "cmsis_gcc.h"`

In this part the the armclang.h line is highlighted as inactive, although the whole project compiles perfectly with it. Perhaps this is also just a bug/missing feature of the cpp-tools extension to be able to use the armclang compiler.

**Comment by aelray** [comment]
Nevermind, I just saw that a fix is underway [here](https://github.com/microsoft/vscode-cmake-tools/issues/1155).

When will that be released?

**Comment by andreeis** [comment]
CMake Tools 1.4.0 Beta is available on GitHub and contains this fix that you are interested in (but which does not address this current issue): https://github.com/microsoft/vscode-cmake-tools/issues/1155. 

https://github.com/microsoft/vscode-cmake-tools/releases/tag/1.4.0-beta

Please try it out and let us know if you encounter any issues.

**Comment by edgomez** [comment]
Tested newer cmake tools a few times since then.

The problem still persists and the cmake plugin doesn't seem to communicate the C/CXX compiler information  to intellisense w/ the toolchain files with regular variables. The same workarounds are still to be applied: either put the vars in the cache, or manually edit again the toolchain file.

Will there be progress on the proposed solution ? newer cmake propose a file API that can allow the plugin to extract all sorts of informations about the project as well. Pretty sure the C/CXX compiler information is one of them.

**Comment by andreeis** [comment]
Did you try the recent release of CMake Tools? 1.7 contains support for the CMake toolchains object and we get the C/CXX compiler information from there. Let us know if it doesn't fully work for your scenario.

**Comment by edgomez** [comment]
hi @andreeis ,

Yes i'm activelly using the latest vscode + C/C++ insider + cmake tools 1.7.1

For testing, I just created a debian container + mingw installed + handwritten toolchain file.
I compiled a cmake 3.20.2 in the container so that cmake tools had everything required to introspect the fileApi toolchain output.

I noted some improvements. I think we're almost there...

I could find the fileApi toolchain info, showing that cmake tools did request the information, however intellisense remained totally wrong w.r.t defined symbols.

I could trace this to the fact the c_cpp_properties.json configuration entry isn't flexible enough to point it to the correct compile_commands automatically, i had to manually add:
`        {
            "name": "mingw64",
            "includePath": [
                "${workspaceFolder}/**"
            ],
            "compileCommands": "${workspaceFolder}/build/mingw64/compile_commands.json",
            "configurationProvider": "ms-vscode.cmake-tool"
        }
`

I wish we had a ${command:cmake.buildKit} so that i could rewrite this like:
`        {
            "name": "Linux Host based Compilation",
            "includePath": [
                "${workspaceFolder}/**"
            ],
            "compileCommands": "${workspaceFolder}/build/${command:cmake.buildKit}/compile_commands.json",
            "configurationProvider": "ms-vscode.cmake-tool"
        }
`
I would not have to toggle both the build kit and the C/C++ configuration manually in the status bar to get things working.

Note that I didn't check w/ a different CPU target toolchain, so i can't figure if ms-vscode.cmake-tool provides the information if the compiler being used is gcc-x64 or gcc-arm.

**Comment by edgomez** [comment]
Mentioning  #993 which should deliver something close to what is need in my case, but for c_cpp_properties.json

**Comment by andreeis** [comment]
@edgomez, compileCommands is not even used when configuration provider is CMake Tools. I was surprised that adding it fixed IntelliSense for you and then I noticed that there is a typo in the configuration provider id. It's "ms-vscode.cmake-tools". Let us know how things work after you correct that.

**Comment by edgomez** [comment]
@andreeis Good catch !

Indeed, once the typo was fixed, i got the following output in the log:
`[5/7/2021, 6:59:26 PM] For C++ source files, the cppStandard was changed from "" to "gnu++14" based on compiler args and querying compilerPath: "/usr/bin/x86_64-w64-mingw32-g++-posix"
`
So now, I should find out how that typo got there in the first place.

I'll keep the issue opened until i validate a couple of my vendor's ARMv7/AArch64 toolchains. If those succeed, then it'll be time to let go of this issue :-)

thanks !

**Comment by edgomez** [comment]
@andreeis , Well no luck :-(

Tested with an android based container.

I used the android toolchain provided by google within the NDK (`${env:NDK_ROOT}/build/cmake/android.toolchain.cmake`). The C/C++ extension picked up the change and I could see:
```
[5/7/2021, 9:34:52 PM] For C++ source files, IntelliSenseMode was changed from "linux-gcc-x64" to "linux-clang-x64" and cppStandard changed from "" to "c++14" based on compiler args and querying compilerPath: "/opt/android-sdk/ndk/21.3.6528147/toolchains/llvm/prebuilt/linux-x86_64/bin/clang++"
```
That was a great start, but  intellisense still figured it out wrong, notice the x64 arch :-(

So the toolchain information extracted from the fileAPI request provided intellisense/cmake-tools with the correct information for the compiler id Clang. But sadly, intellisense/cmake-tools didn't figure out the target arch aarch64. This is not so surprising for clang, as it's a cross compiler by design. Running clang to check the target arch is meaningless, unless the correct options are applied.

Btw here are the cmake configuration args i used to compile the sources in `settings.json`.
```
   "cmake.configureArgs": [
        "-DANDROID_ABI=arm64-v8a",
        "-DANDROID_PLATFORM=android-24",
   ...
  ]
```
In the resulting cmake 3.20 toolchain fileApi file, i could see the fields:
1.  `"id" : "Clang"`
2.  `"target" : "aarch64-none-linux-android24"`

That should be plenty for intellisense to figure out it's working with a Clang compiler (done) targeting aarch64+linux android.

Am i missing something ?

Again if we need these informations even w/o the fileApi toolchain file from cmake >= 3.20, i can provide cmake code that the cmake-tools plugin could configure, and then read back the relevant CMake variable values it needs. That solution would have the benefit to  be compatible w/ cmake 3.x (x to be determined, but surely <= 13 as this is what i use daily). I don't have the required knowledge to hack my way in the C/CPP or cmake-tools VSCode plugins to try to integrate my proposal. But i'm willing to spend some time providing a cmake script that can help extract any required CMake variable for cmake-tools/intellisense to work out of the box.

The cmake script could extract CMAKE_SYSTEM_NAME, CMAKE_SYSTEM_PROCESSOR, CMAKE_C_COMPILER, CMAKE_CXX_COMPILER as a start

**Comment by andreeis** [comment]
@edgomez, thank you for the last comments, I will read them more thoroughly. For now I have a question: do you confirm that for CMake >= 3.20, all works well? This report we can use to track how to fix this for < 3.20? Or >= 3.20 still has issues?

**Comment by edgomez** [comment]
@andreeis : for cmake >= 3.20, vscode does not account for all toolchain fileApi information. It found it was clang, but didn't check which target architecture. The test was based on a docker based android toolchain. I can provide the Dockerfile if needed.

**Comment by andreeis** [comment]
adding @benmcmorran, for the >= 3.20 part of this issue.

**Comment by andreeis** [comment]
@benmcmorran looked at the >=3.20 scenario and developed a fix here https://github.com/benmcmorran/vscode-cmake-tools/commit/1619ffcddebe35e5024e12b6b03e693b2e7f9a6a, that we will pickup and finalize (needs one more fix). I am opening a new work item for this and I am leaving this report open for the < 3.20 scenario and the suggestion from @edgomez.

**Comment by andreeis** [comment]
@edgomez and @aelray, can you try the CMake Tools vsix from [here](https://github.com/microsoft/vscode-cmake-tools/actions/runs/920800165) and confirm if it solves anything for you, when you use a CMake at least as new as 3.20?

**Comment by aelray** [comment]
@andreeis should this build help with regard to #1877? I just tested it and it seems to still not provide the correct Intellisense configuration.

**Comment by andreeis** [comment]
No, not that one, but for latest comments from @edgomez. I think you were referring to different issues earlier in this report (it became too large and with multiple different problems being described :), so a bit confusing). So maybe you can ignore the vsix and I'll wait for @edgomez answer.

**Comment by edgomez** [comment]
@andreeis ok, i think the best approach to tackle this is simply to provide you with a test environment. Once setup you should have everything ready for tracking the issue and fixing it.

https://github.com/edgomez/cmake-tools-1188-env

Please follow the README.md instructions and let me know if you run into trouble using the environement.

**Comment by bilderbuchi** [comment]
Related: #1278 I think.
